### PR TITLE
Fix batch updates to allow cleared values to be re-set

### LINKED ISSUE #5697: [master branch] Theme switching in Control Catalog is broken
**Describe the bug**
Application crashes, when user switches theme.

**To Reproduce**

1. Open control catalog
2. Select any other theme in the themes drop down
3. Wait couple of seconds

**Desktop (please complete the following information):**

- OS: Windows, Mac (I assume any other OS affected too)
- Version: master branch https://github.com/AvaloniaUI/Avalonia/commit/8bede23d7bdbf87ec052ea00e166744382aa83ec

**Additional context**

Probably introduced by https://github.com/AvaloniaUI/Avalonia/pull/5070
**IS NOT** fixed with this PR - https://github.com/AvaloniaUI/Avalonia/pull/5686

```
An item with the same key has already been added. Key: Avalonia.Animation.DoubleTransition
   at System.ThrowHelper.ThrowAddingDuplicateWithKeyArgumentException[T](T key)
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at Avalonia.Animation.Animatable.AddTransitions(IList items) in line 189
```

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

@MarchingCube reported a problem introduced by #5070 whereby if a property was cleared during a batch update and then re-set in the `PropertyChanged` notification raised on ending the batch update, an exception would be thrown.

This was caused by our usage of sentinel values to indicate that a local value property entry needs to be removed at the end of a batch update. The fixes are:

- If a new local value is written to the store in the course of ending the batch update, don't update the sentinel value as the value will subsequently be lost. Instead create a new local value
- Allow adding a binding to a cleared property while ending a batch update
- When a binding is added during the end of a batch update, `_batchUpdate` will be non-null but newly added bindings shouldn't have `BeginBatchUpdate` called on them because no `EndBatchUpdate` will arrive (as we've already called them)

Update: it seems #5697 was also caused by batch updating. The `ValueStore` fix is:

- Fix bindings being disposed during batch update

But it seems that batch updates also exposed some problems with transitions. Fix those:

- Handle changing a `Transitions` collections to new collection which contains some of the same elements
- Handle changing transitions during batch update: transition state may not have been initialized by the time the change notification is received for the property which has a transition

Fixes #5697

### CODE REVIEW

File: src/Avalonia.Base/ValueStore.cs [hunk 1]
```diff
@@ -103,7 +103,7 @@ public bool IsSet(AvaloniaProperty property)
 
             IDisposable? result = null;
 
-            if (_values.TryGetValue(property, out var slot))
+            if (_values.TryGetValue(property, out var slot) && !IsRemoveSentinel<T>(slot))
```
Comment (MarchingCube) [inline src/Avalonia.Base/ValueStore.cs:106]: I'm still not 100% sure why only certain places respect sentinel values, perhaps we should introduce a small helper to get values (with or without sentinel value handling). And perhaps a comment explaining why we need to handle sentinel values in certain cases would help with maintenance.

**Comment by MarchingCube** [comment]
I've asked @Michron to check if it fixes problems on our side.

**Review by MarchingCube — APPROVED** [review]
Fix confirmed.

File: src/Avalonia.Base/ValueStore.cs (hunk 1 above)
Comment (grokys) [reply src/Avalonia.Base/ValueStore.cs]: Good point, I've made handling sentinel values more consistent.

**Comment by grokys** [comment]
@MarchingCube should now also fix #5697.

File: src/Avalonia.Base/ValueStore.cs [hunk 2]
```diff
@@ -437,9 +456,14 @@ public bool End()
...
                             // During batch update values can't be removed immediately because they're needed to raise
                             // the _sink.ValueChanged notification. They instead mark themselves for removal by setting
-                            // their priority to Unset.
-                            if (slot.Priority == BindingPriority.Unset)
+                            // their priority to Unset. We need to re-read the slot here because raising ValueChanged
+                            // could have caused it to be updated.
+                            if (values.TryGetValue(entry.property, out var updatedSlot) &&
+                                updatedSlot.Priority == BindingPriority.Unset)
                             {
+                                if (entry.property.Name == "Transitions" && _owner._owner.GetType().Name == "TabItem")
```
Comment (MarchingCube) [inline src/Avalonia.Base/ValueStore.cs:464]: Debugging leftover?

Comment (grokys) [reply src/Avalonia.Base/ValueStore.cs:464]: Yep :) removed.

**Comment by MarchingCube** [comment]
Tested again, works fine and fixes both issues 👍

**Pull request merged**